docs/blueprints/ge_refactoring.md

```diff
@@ -0,0 +1,67 @@
+Hello!
+
+This article is all about our vision how to make graceful exit code a bit readable and testable.
+
+So, main logic is placed in Worker (storj/storagenode/gracefulexit/worker.go)
+
+This module has giant list of dependencies, goroutines, large methods, lack of comments and tests.
+
+Moreover, in this place we combine all this logic - database calls, business data processing, different transport level dialings, infinite loops, goroutines.
+
+Size of this code is too large to understand it easily. Sometimes we faced such things like missing returns and wrong error handling.
+
+All this factors won't allow us to be sure that it works fine even if there won't be any bug tickets for some time.
+
+We all understand that small pieces of code are much more readable and its easier to test it and to find potential bug just by reading it.
+
+So, our vision is next:
+
+1. **Worker** - its main responsibility to run whole process. It should run goroutines, listen messages, make all needed dialing.
+It should not depend on some concrete implementation, have references to database, etc.
+This worker should have dependencies interfaces, to be able to mock them in future.
+2. **Service** - service should gather data from database, process them or save them in database.
```

Comment:
I think the service itself shouldnt do any processing, just handing out the DB and static assets like the trust pool and global defaults.

Reply:
I don't see a problem with the service doing processing. I think we are doing this already, and it seems like an appropriate thing to do given our docs: https://github.com/storj/docs/blob/master/code/Service.md

"Services handle internal peer logic. Services use databases or other services to fulfill their logic." - if a service only uses databases or other services to fulfill their logic, they need to have additional processing at some level

Reply:
i think that processing of gathered from DB data - is a part of business logic.
We take data, manipulate it somehow and pass it to next layers\programming modules.

Maybe processing is a bad work for describing this process.

Reply:
maybe we can be more specific and instead of saying "processing", say something like "service should gather data from database, store data in the database, and handle piece transfers between nodes"

Reply:
Yeah thats a good description.